### PR TITLE
gnrc/mac : Expose configurations to Kconfig

### PR DESCRIPTION
**Pull request by akshaim**
### Contribution description

This PR exposes compile configurations in GNRC: MAC to Kconfig.

### Testing procedure

1. New documentation was built using Doxygen 

The build works fine.

2. Macro was introduced 

```
#define STR(x)   #x
#define SHOW_DEFINE(x) printf("%s=%s\n", #x, STR(x))

```
##### Firmware Output

RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2020.07-devel-739-g6e24-Kconfig_mac_tests)
CONFIG_GNRC_MAC_RX_QUEUE_SIZE_EXP=(3U)
GNRC_MAC_RX_QUEUE_SIZE=(1 << (3U))
CONFIG_GNRC_MAC_DISPATCH_BUFFER_SIZE_EXP=(3U)
GNRC_MAC_DISPATCH_BUFFER_SIZE=(1 << (3U))
CONFIG_GNRC_MAC_NEIGHBOR_COUNT=4
CONFIG_GNRC_MAC_TX_QUEUE_SIZE_EXP=(3U)
GNRC_MAC_TX_QUEUE_SIZE=4
CONFIG_GNRC_MAC_DISABLE_DUTYCYCLE_RECORD=CONFIG_GNRC_MAC_DISABLE_DUTYCYCLE_RECORD
GNRC_MAC_ENABLE_DUTYCYCLE_RECORD=(1U)

#### Usage with menuconfig [default values]

> make menuconfig

##### Firmware Output
RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2020.07-devel-739-g6e24-Kconfig_mac_tests)
CONFIG_GNRC_MAC_RX_QUEUE_SIZE_EXP=3
GNRC_MAC_RX_QUEUE_SIZE=(1 << 3)
CONFIG_GNRC_MAC_DISPATCH_BUFFER_SIZE_EXP=3
GNRC_MAC_DISPATCH_BUFFER_SIZE=(1 << 3)
CONFIG_GNRC_MAC_NEIGHBOR_COUNT=8
CONFIG_GNRC_MAC_TX_QUEUE_SIZE_EXP=3
GNRC_MAC_TX_QUEUE_SIZE=(1 << 3)
CONFIG_GNRC_MAC_DISABLE_DUTYCYCLE_RECORD=CONFIG_GNRC_MAC_DISABLE_DUTYCYCLE_RECORD
GNRC_MAC_ENABLE_DUTYCYCLE_RECORD=(1U)

#### Usage with menuconfig

> make menuconfig

##### Firmware Output
RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2020.07-devel-739-g6e24-Kconfig_mac_tests)
CONFIG_GNRC_MAC_RX_QUEUE_SIZE_EXP=4
GNRC_MAC_RX_QUEUE_SIZE=(1 << 4)
CONFIG_GNRC_MAC_DISPATCH_BUFFER_SIZE_EXP=2
GNRC_MAC_DISPATCH_BUFFER_SIZE=(1 << 2)
CONFIG_GNRC_MAC_NEIGHBOR_COUNT=4
CONFIG_GNRC_MAC_TX_QUEUE_SIZE_EXP=2
GNRC_MAC_TX_QUEUE_SIZE=(1 << 2)
CONFIG_GNRC_MAC_DISABLE_DUTYCYCLE_RECORD=1
GNRC_MAC_ENABLE_DUTYCYCLE_RECORD=(0)

**MACROS were successfully configured.**

### Issues/PRs references

#12888

### CODE REVIEW

**Comment by akshaim** [comment]
@miri64  Do you think that [GNRC_MAC_TYPE_GET_DUTYCYCLE](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/gnrc/mac/types.h#L46-L48) should be made configurable as well ?

**Comment by miri64** [comment]
> @miri64 Do you think that [GNRC_MAC_TYPE_GET_DUTYCYCLE](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/gnrc/mac/types.h#L46-L48) should be made configurable as well ?

No, its an IPC message type. It is unclear to me, why it is configurable in the current version in the first place.

**Comment by akshaim** [comment]
> > @miri64 Do you think that [GNRC_MAC_TYPE_GET_DUTYCYCLE](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/gnrc/mac/types.h#L46-L48) should be made configurable as well ?
> 
> No, its an IPC message type. It is unclear to me, why it is configurable in the current version in the first place.

If you think so, maybe I should remove the `ifndef` statements associated with that macro.

**Comment by miri64** [comment]
> > > @miri64 Do you think that [GNRC_MAC_TYPE_GET_DUTYCYCLE](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/gnrc/mac/types.h#L46-L48) should be made configurable as well ?
> > 
> > 
> > No, its an IPC message type. It is unclear to me, why it is configurable in the current version in the first place.
> 
> If you think so, maybe I should remove the `ifndef` statements associated with that macro.

Yes, but this is IMHO a separate change from the Kconfig integration.

**Comment by akshaim** [comment]
> > > > @miri64 Do you think that [GNRC_MAC_TYPE_GET_DUTYCYCLE](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/gnrc/mac/types.h#L46-L48) should be made configurable as well ?
> > > 
> > > 
> > > No, its an IPC message type. It is unclear to me, why it is configurable in the current version in the first place.
> > 
> > 
> > If you think so, maybe I should remove the `ifndef` statements associated with that macro.
> 
> Yes, but this is IMHO a separate change from the Kconfig integration.

Okay. fine. I shall make a separate PR.

**Comment by akshaim** [comment]
@miri64  Okay. I shall go for a separate PR then.

File: sys/net/gnrc/link_layer/mac/Kconfig [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# Copyright (c) 2020 Freie Universitaet Berlin
+#
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+#
+menuconfig KCONFIG_MODULE_GNRC_MAC
+    bool "Configure GNRC_MAC"
```
Comment (leandrolanzieri) [inline sys/net/gnrc/link_layer/mac/Kconfig:8]:
```suggestion
    bool "Configure GNRC MAC"
```

File: sys/net/gnrc/link_layer/mac/Kconfig [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    depends on MODULE_GNRC_MAC
+    help
+        Configure the GNRC_MAC using Kconfig.
+
+if KCONFIG_MODULE_GNRC_MAC
+
+config GNRC_MAC_RX_QUEUE_SIZE_EXP
+    int "Exponent; RX queue size (resulting in the queue size 2^n)"
```
Comment (leandrolanzieri) [inline sys/net/gnrc/link_layer/mac/Kconfig:16]:
```suggestion
    int "Exponent for the RX queue size (resulting in the queue size 2^n)"
```

File: sys/net/gnrc/link_layer/mac/Kconfig [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    default 3
+    help
+        As the queue size ALWAYS needs to be power of two, this option
+        represents the exponent of 2^n, which will be used as the size of
+        the RX queue for incoming packets.
+
+config GNRC_MAC_DISPATCH_BUFFER_SIZE_EXP
+    int "Exponent; Dispatch buffer size (resulting in the buffer size 2^n)"
```
Comment (leandrolanzieri) [inline sys/net/gnrc/link_layer/mac/Kconfig:24]:
```suggestion
    int "Exponent for the dispatch buffer size (resulting in the buffer size 2^n)"
```

File: sys/net/gnrc/link_layer/mac/Kconfig [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        As the buffer size ALWAYS needs to be power of two, this option
+        represents the exponent of 2^n, which will be used as the size of
+        the RX queue for incoming packets.
+
+config GNRC_MAC_NEIGHBOR_COUNT
+    int "Count of neighbor nodes in one-hop distance"
+    default 8
+
+config GNRC_MAC_TX_QUEUE_SIZE_EXP
+    int "Exponent; TX queue size (resulting in the queue size 2^n)"
```
Comment (leandrolanzieri) [inline sys/net/gnrc/link_layer/mac/Kconfig:36]:
```suggestion
    int "Exponent for the TX queue size (resulting in the queue size 2^n)"
```

**Comment by akshaim** [comment]
This was to reduce t

> Some minor comments. Besides that I'd rather have just one commit that updates the documentation after all changes are done. That would be clearer.

Sure. I shall do that from now on. I used ';' to restrict the line to 80 chars.

**Comment by leandrolanzieri** [comment]
@akshaim please squash, and try merging the documentation commits

**Comment by akshaim** [comment]
@leandrolanzieri Please wait. Need to test.

**Comment by akshaim** [comment]
> @leandrolanzieri Please wait. Need to test.

Done.

**Comment by leandrolanzieri** [comment]
I think something happened when squashing. Why is `GNRC_MAC_DISPATCH_BUFFER_SIZE` being removed in 0921e6ded9ed01dbf1f9f395d7420701a13f6d4b?

**Comment by akshaim** [comment]
> I think something happened when squashing. Why is `GNRC_MAC_DISPATCH_BUFFER_SIZE` being removed in [0921e6d](https://github.com/RIOT-OS/RIOT/commit/0921e6ded9ed01dbf1f9f395d7420701a13f6d4b)?

Thanks for pointing that out. Multiple rebases to merge the documentation resulted in some unexpected changes. The final commit passed the tests however I missed the intermediate ones.

**Pull request merged**